### ISSUE TITLE
Filter using like operator

### ISSUE THREAD

**Issue by npatmaja** [scoutforpets/bookshelf-jsonapi-params#19]
Hi, is the package currently support like operator statement, e.g., `name like %john%`.

Currently, I define the `like` operator as `contains`, so the URL would become like 

```
/users?filter[email][contains]='gmail'
```

Then, I separated out that filter and build `where` statements manually based on that (I build a [bookshelf plugin](https://github.com/npatmaja/bookshelf-flexi-where) for this purpose). It would be really great if it is actually supported by bookshelf-jsonapi-params directly.

**Comment by jamesdixon** [comment]
Hi @npatmaja,

Thanks for your message.

Unfortunately, it doesn't support this type of syntax at the moment. The JSON API spec unfortunately isn't very specific on anything but the simplest of filtering syntax, so I kept things simple. That said, I'd be interested in possibly integrating your plugin to enable this as I've run into this issue myself. More than willing to accept a PR!

Thanks,
James

**Comment by npatmaja** [comment]
@jamesdixon Sorry for the late reply.

First of all, is the syntax ok (using contains, lt, gt, lte, gte)?

I still need to figure out how to best integrate the feature with this module, and maybe #16 as well. Currently my solution was to create a new plugin to extend **bookshelf-jsonapi-params**, but it feels a bit hacky at the moment.

**Comment by alechirsch** [comment]
Hi, I recently just took a stab at adding functionality for filtering by a specific "type".
I added support for 'like', 'not', 'lt', 'gt', 'lte', and 'gte'. Let me know what you think of it.
[https://github.com/alechirsch/bookshelf-jsonapi-params](https://github.com/alechirsch/bookshelf-jsonapi-params)

**Comment by jamesdixon** [comment]
@npatmaja I like that syntax and think it's a good way to go!

@alechirsch it looks like you've implemented exactly what @npatmaja suggested? I didn't review the code in detail, but seems great. Can you write up some tests to go along with it? If so, more than happy to accept a PR.

Thanks, gents!

**Comment by alechirsch** [comment]
I will look into writing some tests as well as updating the README, I should be able to get to it within the next few days.

**Comment by jamesdixon** [comment]
Great. Thank you!

**Comment by alechirsch** [comment]
In addition to my fork of this repository, I also forked node-jsonapi-query-parser to support this feature.
[https://github.com/alechirsch/node-jsonapi-query-parser](https://github.com/alechirsch/node-jsonapi-query-parser)

**Comment by jamesdixon** [comment]
Any chance you can contact the owner of that repo and see about getting your changes merged?

**Comment by alechirsch** [comment]
Yeah, I will look into it

**Comment by npatmaja** [comment]
@alechirsch cool!

I skimmed over the code and I think it can be improved by separating out the checking of the operators (contains/like, lt, gt, etc) so that we can reuse it when doing #16, as the deep relation filter should be able to have this functionality, I assume. Furthermore if we want to add more operator, a _between_ perhaps, it will be easier as the code is isolated in a separate function.

@jamesdixon I guess we need to decide which syntax we need to support as I believe it is not specified in JSON API spec yet.

If I want to collaborate, how should I proceed given @alechirsch has been doing the first iteration, should I create a PR from to @alechirsch's fork first? Sorry I'm still new in collaborating in an open source project. :)

**Comment by alechirsch** [comment]
I think #16 is a completely separate beast that we should consider after this feature is in.

@npatmaja Do you have any suggestions to improve the syntax that I have implemented, we should definitely collaborate on that. I still want to keep it separate from "filter" so we don't affect what is already implemented and being used by other people.

@jamesdixon I will add tests and Documentation for what I implemented either later today or tomorrow. Would you mind taking a closer look at the code and giving me your thoughts on it?

**Comment by jamesdixon** [comment]
@alechirsch thanks for adding tests/docs. I'll do a thorough code review soon.

Regarding the syntax, I actually prefer the syntax that @npatmaja suggested as it doesn't introduce another parameter to deal with. From what I can tell, I don't see it having any conflict for previous user. For example: 

`/pets?filter[name]=max` would return all pets named max. It's a straight equality comparison.

`/pets?filter[weight][gt]=50` would return all pets with a weight greater than 50lbs.

If there isn't an operator, we just assume it's an equality operation. Thoughts?

The tricky part comes when dealing with `and`, `or`, `between`, etc :) We'll leave that one for another day!

**Comment by jamesdixon** [comment]
This might be worth looking at: http://elide.io/pages/guide/08-filters.html

Elide is Yahoo's java implementation of JSON API

**Comment by npatmaja** [comment]
@jamesdixon @alechirsch I think it should be the case for filters without operator, it should be interpreted as _equal_ operator or _in_ if it is supplied with a list of values, as mentioned in http://jsonapi.org/recommendations/#filtering

For `between` I guess we could do something like `?filter[weight][between]=25..10` or `?filter[weight][between]=25,10`. For `and` and `or` it is really tricky, as those two can be used in top level filters or inside a filter, e.g., 

```
select * from pets where owner = 'john doe' and (weight >= 50 or weight <= 20) or age >= 20
```

**Comment by alechirsch** [comment]
It may be tricky to avoid conflicts in the filter object. The filter object would have both strings and objects as fields.  more specifically, it may have an equality field as well as always having the objects like, not, gt, lt, gte, and lt. While it is unlikely that there would be a column name with one of these filter types, its certainly a possibility. 

That was the main reason why I kept the filterType separate from filter.

**Comment by alechirsch** [comment]
It is certainly possible to avoid conflicts and have this syntax /pets?filter[weight][gt]=50&filter[like]=true if we keep the backend structure the same. If a user sends the above query, the result from  node-jsoapi-query-parser it would be the following

```
queryData: {
    include: [],
    fields: {},
    sort: [],
    page: {},
    filter: {
      like: true
    },
    filterType: {
      like: {},
      not: {},
      lt: {},
      lte: {},
      gt: {
          weight: '50'
       },
      gte: {}
    }
  }
```

**Comment by jamesdixon** [comment]
@npatmaja 

> I think it should be the case for filters without operator, it should be interpreted as equal operator or in if it is supplied with a list of values, as mentioned in http://jsonapi.org/recommendations/#filtering

Agreed.

> For between I guess we could do something like ?filter[weight][between]=25..10 or ?filter[weight][between]=25,10

The comma syntax is probably the simplest and easiest to deal with. I'm good with it for now.

@alechirsch I'd prefer to keep the syntax using only a `filter` attribute. Given that JSON API hasn't specified any other fields related to filter (nor anything else related to the implementation for that matter), I think it might be best to work within the parameters that have been given to us rather than adding a completely new parameter. 

I agree that while it's possible that there could be columns with those names, I think you're also correct that it's highly unlikely. I'm also of the mind that the implementation should be as simple as possible. Ultimately, we provide a set of constructs for the most common cases (like, not, gt, between, etc) and if someone needs something more specific or has a conflict with a column name, they have the option to intercept those parameters and write a custom query using Bookshelf + Knex. In that same vein, I was looking at the Elide syntax and WOW...intense and probably overkill IMO.

Thanks for both of your help on this and looking forward to hearing your thoughts!

**Comment by alechirsch** [comment]
Ok, in the documentation we should say that like, not, gt, ... are all reserved keywords. I am not going to add any more keywords at the moment.

@jamesdixon Do you think the syntax should be filter[like][name] or filter[name][like]?

**Comment by jamesdixon** [comment]
> Ok, in the documentation we should say that like, not, gt, ... are all reserved keywords. I am not going to add any more keywords at the moment.

Makes sense!

> Do you think the syntax should be filter[like][name] or filter[name][like]?

I believe the latter makes the most sense given that if it's an equality or in operation, there's no operator and we're specifying the column name. The column name would be the first regardless.

Thanks again, @alechirsch!

**Comment by npatmaja** [comment]
@jamesdixon @alechirsch 👍  cool!

Thanks gentlemen!

**Comment by alechirsch** [comment]
@jamesdixon I implemented the changes to the syntax and updated all of the tests. I also updated node-jsonapi-query-parser with the new syntax changes.

Let me know what you think and if it is ready to be merged.

**Comment by gmarkv10** [comment]
@jamesdixon I am working with an Ember frontend and just came across an issue when trying to dynamically generate filter options from a visual search component that makes a strong case against the [name][like] order.

In our search component you have the option to select an option for equivalency or to use the 'like' operator, here is where I see the problem become ill-defined with [name][like] ordering:

> I have two search criteria-
> - Name: Like: AA
> - Name: Bananas

Semantically I think we can agree this means I want things with AA in their name OR the name Bananas. The problem comes, when making a JS-flavored object for the filtering options like you do for Embers query function. I can make an object like 

```
{
  name: {
    like: AA
  }
}
```

OR an object like 

```
{
  name: Bananas
}
```

But can not simultaneously have that object's 'name' property point to both a string Bananas and an object holding the 'like' criteria. It also seems like when it comes time to use this object on the backend with this module, everything is set up with the [like][name] configuration anyway. The argument that filterType keywords might be column names is valid - but the way you want to expose filters and typed filters in the URL will cause a lot of people to have to modify the adapters they are using on the frontend instead of being able to work with tidy objects like this that will work out of the box.

```
{
  like: {
    name: AA
  },
  name: Bananas
} 

```

Having seen this from a frontend perspective do you still feel [name][like] is best?

**Comment by jamesdixon** [comment]
@gmarkv10 thanks for your input. Based on your example, I can see where this could be a problem.

@alechirsch do you have any input on this?

**Comment by alechirsch** [comment]
@jamesdixon This is not an issue for this specific package. The discussion around the syntax is more related to how node-jsonapi-query-parser works.

I think it should be [like][name] simply because we should keep it consistent with the way we can access a javascript object.

As @gmarkv10  pointed out, we cannot have a filter object that looks like this:

```
{
  name: {
    like: 'AA'
  },
  name: 'Bananas'
} 
```

The filter object that this package expects currently looks like this:

```
{
  like: {
    name: 'AA'
  },
  name: 'Bananas'
} 
```

And with javascript we can access the filter object like this:

```
filter[like][name]
filter[name]
```

While the syntax filter[name][like] does look and sound better, for technical reasons I do not think it is the best approach.

**Comment by jamesdixon** [comment]
Fair enough. Thank you for the clarification

**Comment by jamesdixon** [comment]
Closed via #22
